File: crates/taflow-core/src/stream/cumulative_maximum.rs
//! Persistent cumulative maximum state.

/// Cumulative maximum of scalar observations.
#[derive(Debug, Clone)]
pub struct CumulativeMaximum {
    extreme: f64,
    value: Option<f64>,
}

impl CumulativeMaximum {
    /// Create an empty cumulative maximum.
    pub fn new() -> Self {
        Self { extreme: f64::NEG_INFINITY, value: None }
    }

    /// Append one observation and return the cumulative maximum.
    pub fn append(&mut self, input: f64) -> f64 {
        self.extreme = self.extreme.max(input);
        self.value = Some(self.extreme);
        self.extreme
    }

    /// Return the latest cumulative maximum.
    pub fn value(&self) -> Option<f64> { self.value }

    /// Reset the accumulated maximum.
    pub fn reset(&mut self) {
        self.extreme = f64::NEG_INFINITY;
        self.value = None;
    }
}

impl Default for CumulativeMaximum {
    fn default() -> Self { Self::new() }
}

File: python/taflow/cumulative_maximum.py
"""Persistent cumulative maximum indicator."""

from typing import Any

import numpy as np

from ._native import CumulativeMaximum as _NativeCumulativeMaximum
from ._series import as_float64_series


class CumulativeMaximum:
    """Compute the cumulative maximum in persistent Rust state.

    ``_input`` is a required chronological numeric series; pass an empty array
    for a fresh streaming state. There is no warm-up. Correctness maps to the
    Polars ``Series.cum_max`` expression.
    """

    def __init__(self, _input: Any) -> None:
        self._state = _NativeCumulativeMaximum()
        self.extend(_input)

    def append(self, _input: float) -> "CumulativeMaximum":
        """Append one observation and return this indicator."""
        self._state.append(float(_input))
        return self

    def extend(self, _input: Any) -> "CumulativeMaximum":
        """Append chronological observations and return this indicator."""
        self._state.extend(as_float64_series(_input))
        return self

    def compute(self) -> np.ndarray:
        """Return the complete aligned ``float64`` history."""
        return self._state.compute()

    @property
    def value(self) -> float | None:
        """Return the latest result, or ``None`` before the first value."""
        return self._state.value

    def reset(self) -> "CumulativeMaximum":
        """Restore fresh native state and return this indicator."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of observations recorded by Rust."""
        return len(self._state)


__all__ = ["CumulativeMaximum"]
